### PR TITLE
fix(frontend): Add type check for constant bit shift overflows

### PR DESCRIPTION
**Pull request by vezenovm**
# Description

## Problem\*

Resolves issue https://github.com/noir-lang/noir/issues/2250 which was solved incorrectly previously. Reference the summary for details

## Summary\*

After https://github.com/noir-lang/noir/pull/2481 running the `bit_shifts_comptime` test in debug mode an overflow was being triggered during ACIR generation. This overflow should have been caught earlier, thus I added a case to pass during type checking and added a compiler sanity check in case the bug comes back. 

An example error: 
Screenshot 2023-08-30 at 4 25 54 PM

Runtime overflows are still possible, but these are being handled with (https://github.com/noir-lang/noir/pull/2180). 

## Documentation

- [ ] This PR requires documentation updates when merged.

  

  - [ ] I will submit a noir-lang/docs PR.

  

  - [ ] I will request for and support Dev Rel's help in documenting this PR.

  
  

## Additional Context

# PR Checklist\*

- [X] I have tested the changes locally.
- [X] I have formatted the changes with [Prettier](https://prettier.io/) and/or `cargo fmt` on default settings.

### CODE REVIEW

File: crates/noirc_evaluator/src/ssa/ir/instruction.rs [hunk 1]
```diff
@@ -778,12 +778,10 @@ impl Binary {
...
                 // The divisor is being truncated into the type of the operand, which can potentially
-                // lead to the rhs being zero.
-                // If the rhs of a division is zero, attempting to evaluate the divison will cause a compiler panic.
-                // Thus, we do not evaluate the division in this method, as we want to avoid triggering a panic,
-                // and the operation should be handled by ACIR generation.
+                // lead to the rhs being zero. This should be caught before instruction simplification
+                // and passing this if statement is a compiler bug.
                 if matches!(self.operator, BinaryOp::Div) && rhs == 0 {
-                    return None;
+                    unreachable!("ICE: the divisor of a binary op has been truncated into zero");
```
Comment (jfecher) [inline crates/noirc_evaluator/src/ssa/ir/instruction.rs:784]: I was thinking this was going to trigger if we did `let z: u32 = 3 / 0`, but it does not for some reason. Do you know why?

File: crates/noirc_frontend/src/hir/type_check/stmt.rs [hunk 2]
```diff
@@ -287,13 +288,34 @@ impl<'interner> TypeChecker<'interner> {
...
             }
+            HirExpression::Infix(HirInfixExpression { operator, rhs, .. }) => {
+                if operator.is_bit_shift() {
+                    let rhs_expr = self.interner.expression(&rhs);
+                    let rhs_value: u128 = match rhs_expr {
+                        HirExpression::Literal(HirLiteral::Integer(value)) => value.to_u128(),
+                        _ => return,
+                    };
+                    if let Type::Integer(_, bit_count) = annotated_type {
+                        let max: u128 = 1 << bit_count;
```
Comment (jfecher) [inline crates/noirc_frontend/src/hir/type_check/stmt.rs:306]: This needs to be done in SSA or later. The type checker cannot catch cases like `let z: u8 = 3 >> (4 + 4);`

**Review by jfecher — CHANGES_REQUESTED** [review]

File: crates/noirc_evaluator/src/ssa/ir/instruction.rs (hunk 1 above)
Comment (vezenovm) [reply crates/noirc_evaluator/src/ssa/ir/instruction.rs:784]: `eval_constants` is only [called](https://github.com/noir-lang/noir/blob/0920dd03d67c50da36bfb87db2e50f6a4aa155bd/crates/noirc_evaluator/src/ssa/ir/instruction.rs#L638)  after a check whether the rhs is `0` and we are not attempting a div. The truncate inside of `eval_constants` was truncating a value into the type of the lhs. In the case of `let a: u1 = 1 >> 1` we are attempting to do `1 / 2`, and thus are attempting to truncate 2 into a `u1` which is why the rhs would be 0. For `let z: u32 = 3 / 0` we never hit `eval_constants`.

Comment (jfecher) [reply crates/noirc_evaluator/src/ssa/ir/instruction.rs:784]: I understand that - I was missing that we actually have another of the same check on line 635 that checks if we're dividing with an rhs of zero. That prevents eval_constants from being called. Perhaps we only need one of these two checks.

Comment (vezenovm) [reply crates/noirc_evaluator/src/ssa/ir/instruction.rs:784]: Yeah we should really only need the first one on 635. It was a mistake to originally have this check inside `eval_constants`, but I thought to leave it as a compiler sanity check for a more clear error than the panic shown in #2250. If we deem the check unnecessary I could just leave a comment stating that if we hit a divide by zero panic it is most likely due to a missing type check for an overflow.

**Comment by kevaundray** [comment]
@vezenovm should this be closed?

**Comment by vezenovm** [comment]
Yes this is superceded by https://github.com/noir-lang/noir/pull/2713. Which is blocked at the moment, but this can be closed.

**Pull request closed without merging**